tests/src/certs.c

```diff
@@ -301,13 +301,15 @@ const size_t mbedtls_test_cli_crt_ec_len =
  * Dispatch between SHA-1 and SHA-256
  */
 
-#if defined(PSA_WANT_ALG_SHA_256)
+#if ((MBEDTLS_VERSION_MAJOR == 3) && (MBEDTLS_VERSION_MINOR == 6) \
```

Comment:
Another possibility would be to restore `MBEDTLS_MD_CAN_SHA256` here and in a development test header to have something like:
```C
#if defined(PSA_WANT_ALG_SHA256)
#define MBEDTLS_MD_CAN_SHA256
#endif
```
Wdyt?

Reply:
Other possibility, we keep this file version dependent?

Reply:
I said in the meeting that I though it would be weird to keep this file version-dependent while it's essentially a C version of files that are in the framework. Turns out I was confused: I was thinking about `test_certs.h` not `certs.c`. So, I'm less opposed to keeping this file version-dependent than I was in the meeting.

Still, what I wanted to suggest is that we could have helper macros, but I'd rather give them different names than the 3.6 macros, just for clarity - to make sure we don't accidentally keep using the old macros when we're not supposed to. The macros could be local to this file and be like `CERTS_HAVE_SHA256` etc.

Just a suggestion, now that I've realised the difference between `test_certs.h` and `certs.c` maybe keeping `certs.c` version-dependent is a good option after all.

Reply:
We're likely to need this in several places in the test data. Why would we have different macros in each file?

Regarding this particular file though, do we actually need to prevent unsupported certificates from being defined? This is test data so code size is not really relevant. Why is it bad if we define constants that just don't happen to be used in a particular build?

Reply:
> Regarding this particular file though, do we actually need to prevent unsupported certificates from being defined?

Reply:
> We're likely to need this in several places in the test data. Why would we have different macros in each file?

Indeed. We could have one version-dependent file with a bunch of things like, in development:
```
#if defined(PSA_WANT_ALG_SHA_256)
#define MBEDTLS_TEST_MD_CAN_SHA256
#define MBEDTLS_TEST_SHA256_C
#endif
```
and in 3.6:
```
#if defined(MBEDTLS_MD_CAN_SHA256)
#define MBEDTLS_TEST_MD_CAN_SHA256
#endif
#if defined(MBEDTLS_SHA256_C)
#define MBEDTLS_TEST_SHA256_C
#endif
```
and then everywhere else we just use the `MBEDTLS_TEST_xxx` macros. (Those need to be named after the 3.6 macros, not `PSA_WANT`, otherwise we'd loose information that's useful in 3.6.)

Then we don't even need to explicitly check for the version number anywhere, because we just use those macros and they are defined in a file that's version-dependent so we automagically get the correct definitions (or lack of definition) for the current version.

Wdyt?

Reply:
That seems like a good approach to me, since it solves the essential problem of having version-dependent code in a version-independent repo. This could also work well for the libtestdriver paths:
```c
#define MBEDTLS_TEST_DRIVER_HEADER_DIR "libtestdriver1/tf-psa-crypto/core"
```
and so on.

Reply:
The point of the framework is to share common code. There doesn’t seem to be any commonality in the lines in the PR here - it’s an interleaving of 3.6 and 4.x code, so is really calling to be version-dependent.

`MBEDTLS_TEST_MD_CAN_SHA256 ` &cc seems to be adding an additional layer of indirection: how are we going to test this? Test code needs to be obviously correct, otherwise we need to start testing the test code, then testing the code that tests the test code, then...

(For an example of testing the tests, it's so hard to fathom if we actually do run all the tests that we've written a script `analyze_outcomes.py` to do this for us. That in itself has had bugs, so...)